### PR TITLE
Timing output

### PR DESCRIPTION
**Pull request by mlaures**

### CODE REVIEW

File: gridding.cpp [hunk 1]
```diff
@@ -179,34 +185,61 @@ float computeImageSize(unsigned long grid_size, float end_frequency) {
...
-             idg::Array1D<float> &frequencies,
-             idg::Array1D<std::pair<unsigned int, unsigned int>> &baselines) { // proxy
+             std::shared_ptr<recycle_memory<float>> r_uvw,
+             std::shared_ptr<recycle_memory<float>> r_freq,
+             std::shared_ptr<recycle_memory<std::pair<unsigned int, unsigned int>>> r_base) { // proxy
   
   std::clog << ">>> Initialize IDG proxy." << std::endl;
   idg::proxy::cuda::Generic proxy;
 
+  auto freq = r_freq->operate();
```
Comment (mlaures) [inline gridding.cpp:240]: Frequencies for a single GPU are unlikely to change. Better to keep this outside of the recyclers. Is there another way for this to be managed memory during the timeline of the pipeline?

File: gridding.cpp [hunk 2]
```diff
@@ -237,6 +270,21 @@ void grid_operate_thread(std::shared_ptr<recycle_memory<std::complex<float>>> r3
 
   proxy.init_cache(subgrid_size, cell_size, 0.0, shift);
 
+  auto base = r_base->operate();
```
Comment (mlaures) [inline gridding.cpp:273]: Same for baselines as with frequencies. Will not change order during pipeline

File: gridding.cpp [hunk 3]
```diff
@@ -26,9 +26,10 @@
 
 #include "lender.hpp"
 
-#define TEST false
+#define TEST true
 #define CHANPAR 1 // number of channels to put into one grid
-#define CHANTHR 8 // number of IDG instances to make
+#define CHANTHR 1 // number of IDG instances to make
```
Comment (yupinghuang) [inline gridding.cpp:31]: Add an underscore in there? My first reaction is CHANT_HR. Ditto to CHANPAR especially.

File: gridding.cpp [hunk 4]
```diff
@@ -366,16 +351,15 @@ int main(int argc, char *argv[]) {
...
 
-  // change each of these to recyclers?
+  
+  // This proxy is literally just for these two arrays.
+  idg::proxy::cuda::Generic proxy;
   std::clog << ">>> Allocating metadata arrays" << std::endl;
-  std::vector<size_t> shape_freq {meta.nr_channels};
-  std::shared_ptr<recycle_memory<float>> r_freq = 
-          std::make_shared<recycle_memory<float>>(shape_freq, 1);
+  idg::Array1D<float> frequencies = proxy.allocate_array1d<float>(meta.nr_channels);
```
Comment (yupinghuang) [inline gridding.cpp:358]: We can use your buffer thing to allocate these arrays and then instantiate these array1d instances instead of using proxy. We did that for visibility.

File: gridding.cpp [hunk 5]
```diff
@@ -366,16 +351,15 @@ int main(int argc, char *argv[]) {
...
+  idg::Array1D<std::pair<unsigned int, unsigned int>> baselines =
+      proxy.allocate_array1d<std::pair<unsigned int, unsigned int>>(
+          meta.nr_baselines);
 
-  std::vector<size_t> shape_base {meta.nr_baselines};
-  std::shared_ptr<recycle_memory<std::pair<unsigned int, unsigned int>>> r_base = 
-          std::make_shared<recycle_memory<std::pair<unsigned int, unsigned int>>>(shape_base, 1);
-  
   std::vector<size_t> shape_uvw {meta.nr_baselines, meta.nr_timesteps, 3};
   std::shared_ptr<recycle_memory<float>> r_uvw =
```
Comment (yupinghuang) [inline gridding.cpp:364]: I thought you wrote a buffer_ptr thing that wouldn't require make_shared in the user code?

File: gridding.cpp [hunk 6]
```diff
@@ -186,32 +187,23 @@ float computeImageSize(unsigned long grid_size, float end_frequency) {
 void ms_fill_thread(std::shared_ptr<recycle_memory<std::complex<float>>> r3, 
              const string& ms_path, metadata meta,
              std::shared_ptr<recycle_memory<float>> r_uvw,
-             std::shared_ptr<recycle_memory<float>> r_freq,
-             std::shared_ptr<recycle_memory<std::pair<unsigned int, unsigned int>>> r_base) {
+             idg::Array1D<float> &frequencies,
```
Comment (yupinghuang) [inline gridding.cpp:190]: Why this change?

File: gridding.cpp (hunk 5 above)
Comment (mlaures) [reply gridding.cpp:364]: In this particular case, what's being shared is the recycler... slightly different and unsure if we want another wrapper for that

File: gridding.cpp (hunk 6 above)
Comment (mlaures) [reply gridding.cpp:190]: It doesnt make sense to pass recyclers to the different places for freq/baseline pairs. one thread has to take a buffer, fill it, and then queue it again - and only that one buffer is used by however many threads.

passing object references isnt necessarily "better" but removes the extra steps, because ms_thread changes the object itself and all of the other threads read that

although now that I think about it, the other threads really kind of need to wait for at least that reading bit to be done to properly initialize the proxy...
